src/server/frontend_wayland/foreign_toplevel_list_v1.cpp

```diff
@@ -0,0 +1,560 @@
+-> std::shared_ptr<mw::ExtForeignToplevelListV1::Global>
+{
+    return std::make_shared<ExtForeignToplevelListV1Global>(display, wayland_executor, surface_stack, desktop_file_manager);
+}
+
+// ExtForeignToplevelListV1Global
+
+mf::ExtForeignToplevelListV1Global::ExtForeignToplevelListV1Global(
+    wl_display* display,
+    std::shared_ptr<Executor> const& wayland_executor,
+    std::shared_ptr<SurfaceStack> const& surface_stack,
+    std::shared_ptr<DesktopFileManager> const& desktop_file_manager)
+    : Global{display, Version<1>()},
+      wayland_executor{wayland_executor},
+      surface_stack{surface_stack},
+      desktop_file_manager{desktop_file_manager},
+      id_map{std::make_shared<ForeignToplevelIdentifierMap>()}
+{
+}
+
+void mf::ExtForeignToplevelListV1Global::bind(wl_resource* new_resource)
+{
+    new ExtForeignToplevelListV1{new_resource, *this};
+}
+
+
+// ForeignToplevelIdentifierMap
+
+std::string mf::ForeignToplevelIdentifierMap::toplevel_id(std::shared_ptr<scene::Surface> const& surface)
+{
+    std::lock_guard lock{mutex};
+
+    std::string& identifier = toplevel_ids[surface];
+    if (identifier.empty())
+    {
+        identifier = std::format("toplevel:{:x}", next_id++);
+    }
+    return identifier;
+}
+
+void mf::ForeignToplevelIdentifierMap::forget_toplevel(std::shared_ptr<scene::Surface> const& surface)
+{
+    std::lock_guard lock{mutex};
+    toplevel_ids.erase(surface);
+}
+
+void mf::ForeignToplevelIdentifierMap::forget_stale_toplevels()
+{
+    std::lock_guard lock{mutex};
+    std::erase_if(toplevel_ids, [](auto const& item)
+        {
+            return item.first.expired();
+        });
+}
+
+// ForeignSceneObserver
+
+mf::ForeignSceneObserver::ForeignSceneObserver(
+    std::shared_ptr<Executor> const& wayland_executor,
+    ExtForeignToplevelListV1* manager,
+    std::shared_ptr<DesktopFileManager> const& desktop_file_manager,
+    std::shared_ptr<ForeignToplevelIdentifierMap> const& id_map)
+    : wayland_executor{wayland_executor},
+      manager{manager},
+      desktop_file_manager{desktop_file_manager},
+      id_map{id_map}
+{
+    // Forget toplevel IDs for windows closed while no observer was
+    // running.
+    id_map->forget_stale_toplevels();
+}
+
+mf::ForeignSceneObserver::~ForeignSceneObserver()
+{
+    clear_surface_observers();
+}
+
+void mf::ForeignSceneObserver::surface_added(std::shared_ptr<scene::Surface> const& surface)
+{
+    create_surface_observer(surface);
+}
+
+void mf::ForeignSceneObserver::surface_removed(std::shared_ptr<scene::Surface> const& surface)
+{
+    std::lock_guard lock{mutex};
+    auto const iter = surface_observers.find(surface);
+    if (iter == surface_observers.end())
+    {
+        log_error(
+            "Can not remove ForeignSurfaceObserver: surface %p not in observers map",
+            static_cast<void*>(surface.get()));
+    }
+    else
+    {
+        iter->second->cease_and_desist();
```

Comment:
Aha!

I was wondering why `ForeignSurfaceObserver` contained a mutex - all the observations take place on the `WaylandExecutor`, so there should not need to be any locking. But *this* call gets made on an arbitrary thread.

But that's a problem, because this call will eventually make Wayland calls, and they have to be done on the Wayland thread.

So I think this should be reworked so that `ForeignSurfaceObserver` all happens on the Wayland thread, and then it doesn't need a mutex at all.